### DISCUSSION TITLE
请问如何在中间件中获取GET和POST参数？

### DISCUSSION THREAD

**Discussion by mowocc** [go-kratos/kratos#1925] [Q&A]
1. 请问如何在中间件中获取请求参数，GET和POST参数分别通过什么方式获取，中间件中获取Header参数使用<transport.FromServerContext(ctx)>，但不清楚GET和POST参数获取方式；
2. HTTP请求必须使用application/json类型吗？如果同时兼容其他类型，请问应该如何处理。

**Comment by yzrzr** [comment]
middleware.Handler的第二个参数不就是请求参数么
```go
func MyMiddleware() middleware.Middleware {
	return func(handler middleware.Handler) middleware.Handler {
		return func(ctx context.Context, req interface{}) (interface{}, error) {
			// req 就是参数
			return handler(ctx, req)
		}
	}
}
```
HTTP请求的 Content-Type 没有规定必须要application/json。可以使用什么类型取决于程序启动时是否注册了相应的编解码器，默认支持的类型有 x-www-form-urlencoded、json、proto、xml、yaml 如果想支持其他的类型也很方便，只要实现kratos中的encoding.Codec接口，然后在框架启动时通过encoding.RegisterCodec()方法注册就行

**Reply by xd-fsing** [reply]
能写个例子吗？

**Reply by tureo** [reply]
> 能写个例子吗？

You can refer to this https://github.com/go-kratos/kratos/blob/main/middleware/tracing/span.go#L66-L79
1. Get the transport.Transporter from server context
`tr, ok := transport.FromServerContext(ctx) `
2. Assert tr to http.Transporter
`if ht, ok := tr.(http.Transporter); ok {`
3. Get all standard HTTP request data from`ht.Request()`. Because of `ht.Request()` returns `*http.Request` of the standard library `net/http`
